server/db.go

```diff
@@ -1197,11 +1199,17 @@ func retrieveFromCache(
 	caches map[string]CacheInterface,
 	cacheName string,
 	key string,
+	cohort int16,
 ) interface{} {
 	if caches != nil {
 		if cached, found := caches[cacheName].Get(key); found {
 			return cached
 		}
+		// Ads issuer_type uniqueness relies on a cohort suffix. Check the cache
+		// for the key with the added suffix.
+		if cached, found := caches[cacheName].Get(fmt.Sprintf("%s_%d", key, cohort)); found {
+			return cached
```

Comment:
One bug in the code setting the issuer cache value that means that if the value without a cohort gets set it will cause the issue this PR was raised to address to recur. 

Putting the logic to build the full issuer name at this low level needlessly increases complexity and risk. Would it not be better to centralise the logic around the fact that issuer name has the cohort concatenated within the only function that requires it - `fetchIssuersByCohort`?  If that function started by building the actual issuer name from the type and cohort then used that throughout you would not need to change the cache interface or implementation at all.

Reply:
Agree with this. Feels like `fetchIssuersByCohort` should be responsible for building the key, and this method should remain unchanged.

Reply:
This suffix is only present in the Ads usage of Challenge Bypass. I'm trying to fix this with minimal impact to Grants, which shares this code base and does not have the suffix until we move to using public keys.